index.rst

```diff
@@ -96,11 +102,173 @@ The stated requirement is: Must support a system so that data may be referred to
+input. Attempting to pass a read-only repo as a butler output or a write-only
+repo as a butler input will raise an exception.
+
+
+Already existing Repositories as Outputs
+""""""""""""""""""""""""""""""""""""""""
+
+When initializing a ``Butler`` with one or more output repositories that already
+exist it is important to not add any input repositories that are not already
+parents of the output repository. This is because all butler inputs (including
+readable outputs) become parents of output repositories, and the repository
+configuration is not allowed to change after it has been written.
+
+Output arguments derived from inputs
+""""""""""""""""""""""""""""""""""""
+
+Some settings for output repositories can be derived from input repository
+configurations. For example, if an output configuration does not specify a
+mapper, the input mapper may possibly be assumed (this will work as long as all
+the input repositories use the same type of mapper; if the inputs use different
+types of mapper then a single type mapper can not be inferred to use for the
+output repositories). When possible the butler will use settings from input
+configurations to complete RepositoryArgs parameters for output repositories. 
+
+Search Order
+""""""""""""
+The order of repositories passed to inputs and outputs is meaningful; search is
+depth-first and in order (left to right). Readable outputs will be searched
+before inputs. Parents of readable outputs/inputs will be searched before the
+next passed-in output/input.
+
+Tagging 
+^^^^^^^
+
+Readable repositories can be “tagged” with a non-persisted id that gets used
+when reading from a repository. A tag can be specified in ``RepositoryArgs``
+when initializing a Butler. A repository can be tagged with more than one tag by
+passing in a container of tags. The tag is not persisted with the repository.
+
+When performing read operations on the butler, if the DataId contains one or
+more tags, the repository will only be used for lookups if it is also tagged
+with one of the tags in the DataId. If the DataId has no tags, then all input
+repositories will be used. More information about DataId and its tag are
+available in the DataId section.
+
+RepositoryArgs
+^^^^^^^^^^^^^^
+
+``RepositoryArgs`` instances are used to instantiate repositories in Butler. Its
+parameters are:
+
+* ``mode`` 
+    * Optional.
+    * string - This can be one of 'r', 'w', or 'rw' (read, write, read-write). 
+    * It is used to indicate the read/write state of the repositories. Input
+      repositories are always read-only and an exception will be raised if the
+      mode of an input repository is 'w'. It may be 'rw' but for inputs the
+      behavior will be the same as 'r'. Output repositories must be 'w' or 'rw'.
+      If it is 'rw' the repository will also be used as an input repository. If
+      mode is not specified, outputs will default to 'w' and inputs will default
+      to 'w'.
```

Comment:
Also, I don't understand why 'w' is the default mode for output repositories instead of 'rw'.  /In fact, at least until we actually do switch to a more fine-grained approach with SuperTask in which we're creating new output repositories for every new datasetType, I think we need output repositories to be 'rw'.